src/metadata/metadata.service.ts

```diff
@@ -0,0 +1,79 @@
+import { Injectable, NotFoundException } from '@nestjs/common';
+import {
+  ErrorCode,
+  GlobalMetadataKey,
+  MetadataEntry,
+  SetMetadataParams,
+} from '@polymeshassociation/polymesh-sdk/types';
+import { isPolymeshError } from '@polymeshassociation/polymesh-sdk/utils';
+
+import { AssetsService } from '~/assets/assets.service';
+import { ServiceReturn } from '~/common/utils';
+import { CreateMetadataDto } from '~/metadata/dto/create-metadata.dto';
+import { MetadataParamsDto } from '~/metadata/dto/metadata-params.dto';
+import { SetMetadataDto } from '~/metadata/dto/set-metadata.dto';
+import { PolymeshService } from '~/polymesh/polymesh.service';
+import { TransactionsService } from '~/transactions/transactions.service';
+
+@Injectable()
+export class MetadataService {
+  constructor(
+    private readonly polymeshService: PolymeshService,
+    private readonly transactionsService: TransactionsService,
+    private readonly assetsService: AssetsService
+  ) {}
+
+  public async findGlobalKeys(): Promise<GlobalMetadataKey[]> {
+    return this.polymeshService.polymeshApi.assets.getGlobalMetadataKeys();
+  }
+
+  public async findAll(ticker: string): Promise<MetadataEntry[]> {
+    const { metadata } = await this.assetsService.findOne(ticker);
+
+    return metadata.get();
+  }
+
+  public async findOne({ ticker, type, id }: MetadataParamsDto): Promise<MetadataEntry> {
+    const { metadata } = await this.assetsService.findOne(ticker);
+
+    try {
+      return await metadata.getOne({ type, id });
+    } catch (err: unknown) {
+      if (isPolymeshError(err)) {
+        const { code, message } = err;
+        if (code === ErrorCode.DataUnavailable) {
+          throw new NotFoundException(message);
+        }
+      }
+
+      throw err;
+    }
+  }
+
+  public async create(ticker: string, params: CreateMetadataDto): ServiceReturn<MetadataEntry> {
+    const { signer, webhookUrl, ...rest } = params;
+
+    const {
+      metadata: { register },
+    } = await this.assetsService.findOne(ticker);
+
+    return this.transactionsService.submit(register, rest, {
+      signer,
+      webhookUrl,
+    });
+  }
+
+  public async setValue(
+    params: MetadataParamsDto,
+    body: SetMetadataDto
+  ): ServiceReturn<MetadataEntry> {
+    const { signer, webhookUrl, ...rest } = body;
+
+    const { set } = await this.findOne(params);
+
+    return this.transactionsService.submit(set, rest as SetMetadataParams, {
```

Comment:
The `as SetMetadataParams` makes me a bit nervous, but its not obvious to me how to get rid of it.

If it needs to stay than so be it.

Reply:
This is required in this case. To get rid of it we just need to create a new object (of type SetMetadataParams) with all the keys & values within rest which seems redundant